lib/base/request.js

```diff
@@ -365,20 +366,34 @@ class Request extends EventEmitter {
   }
 
   /**
-   * Sets request to `stream` mode and pulls all rows from all recordsets to a given stream.
+   * Wrap original request in a Readable stream that supports back pressure and pipe to the Writable stream.
+   * It also sets request to `stream` mode and pulls all rows from all recordsets to a given stream.
    *
-   * @param {Stream} stream Stream to pipe data into.
+   * @param {Stream} writableStream Stream to pipe data into.
    * @return {Stream}
    */
-
-  pipe (stream) {
-    this.stream = true
-    this.on('row', stream.write.bind(stream))
-    this.on('done', () => {
-      setImmediate(() => stream.end())
+  pipe (writableStream) {
+    const origRequest = this
+    origRequest.stream = true
+    origRequest.pause()
+    const readableStream = new Readable({
+      objectMode: true,
+      read (/* size */) {
+        origRequest.resume()
+      }
+    })
+    origRequest.on('row', (row) => {
+      if (!readableStream.push(row)) {
+        origRequest.pause()
+      }
+    })
+    origRequest.on('error', (error) => {
+      readableStream.emit('error', error)
+    })
+    origRequest.on('done', () => {
+      readableStream.push(null)
     })
-    stream.emit('pipe', this)
-    return stream
+    return readableStream.pipe(writableStream)
```

Comment:
This will return the `writeableStream` but I think we should return the readable stream... Without that the developer can never have access to the original readable stream we create.

I know this is against the node convention of what `pipe` should return, and maybe we should rename this function so it's not trying to mimic something it isn't.

Reply:
I'm open to whatever suggestions you decide. I didn't expect that they would need to access the intermediate stream since it is being fed from the request. However if you wanted it could be added as a property to the returned value if desired. 

It seemed to me like the pipe function meant to work like this, but just wasn't implemented completely. So I considered this a bug fix. However if you want to keep the original functionality then creating two separate functions is the way to go. I thought for confusion's sake it would be better to just have one pipe that works as the Node.js standard says.